### PR TITLE
Tag backward nodes via _patch_autograd_grad and update remat pass

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #179105
* #179043

Add `_patch_autograd_grad` context manager that patches `autograd.grad`
to install stacktrace preservation hooks (for seq_nr propagation) and
tag backward nodes with `custom["autograd_backward"]` via
`fx_traceback.annotate`.

Update the remat pass to detect backward nodes via two mechanisms:
- `custom["autograd_backward"]`: set automatically by _patch_autograd_grad
- `custom["phase"] == "backward"`: user-level annotation for disambiguation

When the user provides `phase` annotations, only those are used. If
multiple backward regions are detected without user annotation (e.g.
multiple `autograd.grad` calls), the pass errors and asks the user to
annotate the real backward with
`fx.traceback.annotate({"phase": "backward"})`.

Migrate existing tests from the legacy `remat_pass_tag` annotation to
the new `phase: backward` annotation.

Authored with Claude.

ghstack-source-id: 575d00ec6159e4f49250a3a001f146838e9db05b
Pull Request resolved: https://github.com/pytorch/pytorch/pull/179105

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

File: torch/compiler/__init__.py [hunk 1]
```diff
@@ -478,6 +478,47 @@ def _non_strict_tracing_context():
         _is_non_strict_tracing_flag = old
 
 
+@contextlib.contextmanager
+def _patch_autograd_grad():
```
Comment (soulitzer) [inline torch/compiler/__init__.py:482]: If someone stored a reference to .grad somewhere then it wouldn't be patched by this right

Comment (tugsbayasgalan) [reply torch/compiler/__init__.py:482]: Yes... good point. But do we really expect someone do that? I guess we could move this logic inside autograd.grad implementation but that is leaking compiler stuff to eager API.

File: torch/compiler/__init__.py [hunk 2]
```diff
@@ -478,6 +478,47 @@ def _non_strict_tracing_context():
...
+    @functools.wraps(_orig_grad)
+    def _patched_grad(outputs, inputs, *args, **kwargs):
+        if _is_non_strict_tracing():
+            roots = [
+                t.grad_fn
+                for t in (outputs if isinstance(outputs, (list, tuple)) else (outputs,))
+                if isinstance(t, torch.Tensor) and t.grad_fn is not None
+            ]
+            if roots:
+                setup_stacktrace_preservation_hooks(roots)
```
Comment (soulitzer) [inline torch/compiler/__init__.py]: This change is unrelated to this PR right, here all we need is the annotation?

File: torch/compiler/__init__.py [hunk 3]
```diff
@@ -478,6 +478,47 @@ def _non_strict_tracing_context():
...
+    import torch.fx.traceback as fx_traceback
+    from torch._functorch._aot_autograd.logging_utils import (
+        setup_stacktrace_preservation_hooks,
+    )
+
+    _orig_grad = torch.autograd.grad
+
+    @functools.wraps(_orig_grad)
+    def _patched_grad(outputs, inputs, *args, **kwargs):
+        if _is_non_strict_tracing():
```
Comment (soulitzer) [inline torch/compiler/__init__.py]: Aren't we always guaranteed to be non strict tracing, if we patched this.

**Comment by soulitzer** [comment]
Patching sounds fine for now. We can always improve later if necessary.

File: torch/compiler/__init__.py (hunk 3 above)
Comment (soulitzer) [reply torch/compiler/__init__.py]: We could make it an assertion or something instead

File: test/dynamo/test_fx_annotate.py [hunk 4]
```diff
@@ -97,11 +97,15 @@ def fn(x):
...
-            """('call_function', 'sin', {'ac_sin': 0})""",  # noqa: B950
+            """\
+('call_function', 'sin', {'ac_sin': 0})
+('call_function', 'detach_2', {'autograd_backward': True})""",  # noqa: B950
         )
         self.assertExpectedInline(
             str(bw_metadata),
             """\
+('placeholder', 'detach_2', {'autograd_backward': True})
+('call_function', 'sigmoid_backward', {'autograd_backward': True})
```
Comment (soulitzer) [inline test/dynamo/test_fx_annotate.py]: Where did the extra nodes come from

File: torch/compiler/__init__.py (hunk 3 above)
Comment (tugsbayasgalan) [reply torch/compiler/__init__.py]: We also patch in torch.compile path as well. This is for supporting fwd+loss+bwd in compile path.

File: torch/compiler/__init__.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/compiler/__init__.py]: yep not needed, but we do need it for non-strict tracer in general tho. (the non_strict version in torchtitan does this and we plan to land it in pytorch/pytorch)

File: test/dynamo/test_fx_annotate.py (hunk 4 above)
Comment (tugsbayasgalan) [reply test/dynamo/test_fx_annotate.py]: This is because we are patching autograd.grad in compile path as well.

File: test/dynamo/test_fwd_loss_bwd.py [hunk 5]
```diff
@@ -89,6 +89,7 @@ def forward(self, L_mod_parameters_weight_: "f32[4, 4]", L_mod_parameters_bias_:
         loss: "f32[]" = res.sum();  res = None
 
         grad = torch.autograd.grad(loss, (l_mod_parameters_weight_, l_mod_parameters_bias_));  l_mod_parameters_weight_ = l_mod_parameters_bias_ = None
+
```
Comment (soulitzer) [inline test/dynamo/test_fwd_loss_bwd.py:92]: What led to the spacing changes lol

File: torch/compiler/__init__.py (hunk 3 above)
Comment (soulitzer) [reply torch/compiler/__init__.py]: Hmm but in the torch.compile path I understand that to not count as non-strict tracing since dynamo runs?

Comment (soulitzer) [reply torch/compiler/__init__.py]: btw since claude can do this easily - for ease of review and reading, I think it could be good to split out changes as a second PR

File: test/dynamo/test_fwd_loss_bwd.py (hunk 5 above)
Comment (tugsbayasgalan) [reply test/dynamo/test_fwd_loss_bwd.py:92]: It’s just FX readable-printer churn from the new node metadata. After torch.autograd.gradgets tagged with node.meta["custom"]["autograd_backward"], the printer emits an extra blank line after that statement.

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 6]
```diff
@@ -34,27 +34,83 @@ def is_impure_node_for_dce(node: fx.Node) -> bool:
...
+    Duplicate recompute nodes for backward use. DCE removes unused forward versions.
+
+    Backward nodes are identified by custom["phase"] == "backward" (user annotation)
+    or custom["autograd_backward"] == True (set automatically when Dynamo traces
+    torch.autograd.grad).
+    When the user provides phase annotations, only those are used.
+
+    Only a single contiguous backward region is supported. If multiple backward
+    regions are detected (whether from multiple autograd.grad calls or multiple
+    user annotations), an error is raised.
```
Comment (soulitzer) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Technically we could have multiple backward calls and not be detected if they are called consecutively. Maybe that is fine but we could call it out.

Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: yep clarified in the doc

File: torch/compiler/__init__.py (hunk 3 above)
Comment (tugsbayasgalan) [reply torch/compiler/__init__.py]: yep now we just assert here.

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 7]
```diff
@@ -107,7 +164,12 @@ def _gather(n: fx.Node) -> None:
         return deps
 
     # Insert backward nodes
+    # When use_phase_only, only process nodes explicitly tagged with phase: backward.
+    # Otherwise, treat everything from bwd_start onwards as backward.
     for node in list(gm.graph.nodes)[bwd_start:]:
+        if use_phase_only and node not in bwd_node_set:
+            env[node] = new_graph.node_copy(node, lambda x: env[x])
+            continue
```
Comment (soulitzer) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Should we unify this annotation / phase logic with the bwd_start concept rather than have extra branches.

Comment (soulitzer) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Like I expected all the remaining logic to remain unchanged except we now potentially determine bwd_start in more ways.

Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: yep i consolidated now.

File: torch/compiler/__init__.py (hunk 1 above)
Comment (aorenste) [reply torch/compiler/__init__.py:482]: Keep in mind that storing a reference is as simple as:
```
from torch.autograd import grad
```
A better solution which avoids the leaking concerns would be to extend torch.autograd.grad to allow a generic "hook" which can be patched. Something like:
```
_GRAD_HOOK = contextlib.nullcontext()

def grad(...):
  with _GRAD_HOOK:
    ... <the old grad behavior> ...
```

there's be some extra work to preserve and call the old hook in-line when you patch it - but it's a safer solution with the same result and no leaking of compiler into autograd.

File: torch/compiler/__init__.py [hunk 8]
```diff
@@ -478,6 +478,39 @@ def _non_strict_tracing_context():
...
+
+    _orig_grad = torch.autograd.grad
+
+    @functools.wraps(_orig_grad)
+    def _patched_grad(outputs, inputs, *args, **kwargs):
+        if not _is_non_strict_tracing():
+            raise AssertionError(
+                "_patch_autograd_grad() must be used under "
+                "_non_strict_tracing_context()"
+            )
```
Comment (aorenste) [inline torch/compiler/__init__.py:502]: What's the advantage of asserting instead of just calling _orig_grad without the annotation?  i.e.:
```
if _is_non_strict_tracing():
  with fx.traceback.annotate(...):
    return _orig_grad(...)
return _orig_grad(...)
```

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 9]
```diff
@@ -34,31 +34,88 @@ def is_impure_node_for_dce(node: fx.Node) -> bool:
...
-    """Check if node is in backward region via annotation"""
-    return node.meta.get("custom", {}).get("remat_pass_tag", None) == "is_backward"
+def _is_backward_node(node: fx.Node, use_phase_only: bool = False) -> bool:
+    """Check if node is in backward region.
+
+    If use_phase_only is True, only checks custom["phase"] == "backward"
+    (user annotation). Otherwise falls back to custom["autograd_backward"],
+    which Dynamo adds when tracing torch.autograd.grad.
+    """
+    custom = node.meta.get("custom", {})
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: nit: Use a global for {} so you don't allocate an empty dict on every call (here and below)

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 10]
```diff
@@ -34,31 +34,88 @@ def is_impure_node_for_dce(node: fx.Node) -> bool:
     return node.is_impure(impure_random)
 
 
-def _is_backward_node(node: fx.Node) -> bool:
-    """Check if node is in backward region via annotation"""
-    return node.meta.get("custom", {}).get("remat_pass_tag", None) == "is_backward"
+def _is_backward_node(node: fx.Node, use_phase_only: bool = False) -> bool:
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: nit: The name `use_phase_only` implies that when it's False you'll ALSO use phase, not that it's an either-or. Consider `use_phase` (here and below) or maybe an enum.

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 11]
```diff
@@ -34,31 +34,88 @@ def is_impure_node_for_dce(node: fx.Node) -> bool:
...
+    in_backward = False
+
+    for idx, node in enumerate(gm.graph.nodes):
+        is_bwd = _is_backward_node(node, use_phase_only=use_phase_only)
+        if is_bwd:
+            bwd_node_set.add(node)
+            if bwd_start is None:
+                bwd_start = idx
+        if is_bwd and not in_backward:
+            num_regions += 1
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: nit:

```suggestion
            if not in_backward:
                num_regions += 1
```

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 12]
```diff
@@ -108,6 +165,9 @@ def _gather(n: fx.Node) -> None:
 
     # Insert backward nodes
     for node in list(gm.graph.nodes)[bwd_start:]:
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: nit: not from your change but...
```suggestion
    for node in itertools.islice(gm.graph.nodes, bwd_start, None):
```

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: nit: not from your change but:

```suggestion
    for node in itertools.islice(gm.graph.nodes, 0, bwd_start):
```

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 13]
```diff
@@ -108,6 +165,9 @@ def _gather(n: fx.Node) -> None:
...
+        if node not in bwd_node_set:
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Since you only allow a single region, couldn't you change the for loop to:
```
for node in islice(gm.graph.nodes, bwd_start, bwd_start + len(bwd_node_set)):
   ...
```
skip this bwd_node_set check, and then after this loop rip through the remaining nodes like:
```
for node in islice(gm.graph.nodes, bwd_start + len(bwd_node_set)):
    env[node] = new_graph.node_copy(node, lambda x: env[x])
```

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 14]
```diff
@@ -34,31 +34,88 @@ def is_impure_node_for_dce(node: fx.Node) -> bool:
...
-    bwd_start: int | None = None
-    order = {}
-    for idx, node in enumerate(gm.graph.nodes):
-        order[node] = idx
-        if _is_backward_node(node) and bwd_start is None:
-            bwd_start = idx
+    use_phase_only, bwd_start, bwd_node_set, num_regions = (
+        _collect_backward_region_data(gm)
+    )
+    if num_regions > 1:
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py:100]: I wonder if this region determination is too fragile. It's using the node ordering to determine a contiguous region rather than deps. I don't think the graph is guaranteed to be in any particular order other than deps-first so you could easily have a completely valid graph that has a non-backward node in the middle of two backward sections which the backward sections have no deps on (so a simple valid node reordering would make this then pass).

Maybe in practice it's moot.

**Review by aorenste — APPROVED** [review]
None of my comments are blockers although the monkey-patch bothers me. It would be really easy to accidentally foul it.

Approving w/ comments

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py (hunk 13 above)
Comment (soulitzer) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Why do we need to condition on bwd_node_set if we already have bwd_start?

Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: I think the assumption here is that we will have a range of backward nodes, and could have some non-backward nodes right after that

Comment (soulitzer) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Do we expect to support use cases where there will be many nodes after the backward? But also... I wonder if considering them part of backward for the purpose of remat actually bad?

Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: I don't think there will be any at the moment since in the chunked loss part, you would only do it over the backward region. At most there will be some detach nodes i guess. But i do think being principled here and applying only on backward annotated region is probably better even if it makes not much difference

Comment (soulitzer) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: ok ok, but as a simplification, we could do it with bwd_end since we also know the region is contiguous

Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Yeah sure we can do that.

Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: done
